core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java

```diff
@@ -111,65 +146,295 @@ public GeoDistanceRangeQueryBuilder includeLower(boolean includeLower) {
+            }
+            if (point.lon() > 180.0 || point.lon() < -180) {
+                throw new QueryParsingException(context.parseContext(), "illegal longitude value [{}] for [{}]", point.lon(), NAME);
+            }
+        }
+
+        if (coerce) {
+            GeoUtils.normalizePoint(point, coerce, coerce);
+        }
+
+        Double fromValue = null;
+        Double toValue = null;
+        if (from != null) {
+            if (from instanceof Number) {
+                fromValue = unit.toMeters(((Number) from).doubleValue());
+            } else {
+                fromValue = DistanceUnit.parse((String) from, unit, DistanceUnit.DEFAULT);
+            }
+            fromValue = geoDistance.normalize(fromValue, DistanceUnit.DEFAULT);
+        }
+        if (to != null) {
+            if (to instanceof Number) {
+                toValue = unit.toMeters(((Number) to).doubleValue());
+            } else {
+                toValue = DistanceUnit.parse((String) to, unit, DistanceUnit.DEFAULT);
+            }
+            toValue = geoDistance.normalize(toValue, DistanceUnit.DEFAULT);
+        }
+
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new QueryParsingException(context.parseContext(), "failed to find geo_point field [" + fieldName + "]");
+        }
+        if (!(fieldType instanceof GeoPointFieldMapper.GeoPointFieldType)) {
+            throw new QueryParsingException(context.parseContext(), "field [" + fieldName + "] is not a geo_point field");
+        }
+        GeoPointFieldMapper.GeoPointFieldType geoFieldType = ((GeoPointFieldMapper.GeoPointFieldType) fieldType);
+
+        IndexGeoPointFieldData indexFieldData = context.getForField(fieldType);
+        Query query = new GeoDistanceRangeQuery(point, fromValue, toValue, includeLower, includeUpper, geoDistance, geoFieldType,
+                indexFieldData, optimizeBbox);
+        if (queryName != null) {
+            context.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         if (geohash != null) {
-            builder.field(name, geohash);
+            builder.field(fieldName + GeoPointFieldMapper.Names.GEOHASH_SUFFIX, geohash);
         } else {
-            builder.startArray(name).value(lon).value(lat).endArray();
+            builder.startArray(fieldName).value(lon).value(lat).endArray();
+        }
+        builder.field(GeoDistanceRangeQueryParser.FROM_FIELD.getPreferredName(), from);
+        builder.field(GeoDistanceRangeQueryParser.TO_FIELD.getPreferredName(), to);
+        builder.field(GeoDistanceRangeQueryParser.INCLUDE_LOWER_FIELD.getPreferredName(), includeLower);
+        builder.field(GeoDistanceRangeQueryParser.INCLUDE_UPPER_FIELD.getPreferredName(), includeUpper);
```

Comment:
I was about to comment that I don't love the builder referring to the parser, but using ParseField is just better, good one, and it is some piece of info that the two objects have to share indeed.